src/hdmf/backends/hdf5/h5_utils.py

```diff
@@ -494,6 +503,9 @@ def __init__(self, **kwargs):
         call_docval_func(super().__init__, kwargs)
         # Construct the dict with the io args, ignoring all options that were set to None
         self.__iosettings = {k: v for k, v in zip(ioarg_names, ioarg_values) if v is not None}
+        if self.data is None:
+            self.__iosettings['dtype'] = self.dtype
+            self.__iosettings['shape'] = self.shape
```

Comment:
Since ``dtype`` and ``shape`` are only allowed if ``data`` is ``None`` I would suggest adding a check here to raise a ``ValueError`` if the user tries to provide ``data`` and change the ``dtype`` or ``shape``.  We should then also add a unit test to check for those errors. 

```suggestion
        if self.data is None:
            self.__iosettings['dtype'] = self.dtype
            self.__iosettings['shape'] = self.shape
        else:
            if getargs('dtype', kwargs) is not None:
                 raise ValueError("Setting the dtype when data is not None is not supported")
            if  getargs('shape', kwargs) is not None:
                 raise ValueError("Setting the shape when data is not None is not supported")
```

Reply:
@oruebel I moved this to DataIO. What do you think?

Reply:
Sounds reasonable. In that case I would update the doc for those parameters in `DataIO` as well to mention this behavior.